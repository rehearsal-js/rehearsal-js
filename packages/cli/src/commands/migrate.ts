#!/usr/bin/env node

// TODO: handle ctrl + c

import { migrate } from '@rehearsal/migrate';
import {
  discoverEmberPackages,
  getMigrationStrategy,
  SourceFile,
} from '@rehearsal/migration-graph';
import { jsonFormatter, mdFormatter, Reporter, sarifFormatter } from '@rehearsal/reporter';
import { Command } from 'commander';
import { existsSync, readJSONSync, rmSync, writeJsonSync } from 'fs-extra';
import { Listr } from 'listr2';
import { resolve } from 'path';
import { createLogger, format, transports } from 'winston';

import { generateReports } from '../helpers/report';
import {
  MigrateCommandContext,
  MigrateCommandOptions,
  PackageMap,
  PackageSelection,
} from '../types';
import { UserConfig } from '../userConfig';
import { addDep, determineProjectName, parseCommaSeparatedList, runModuleCommand } from '../utils';

function ifHasTypescriptInDevdep(basePath: string): boolean {
  const packageJSONPath = resolve(basePath, 'package.json');
  const packageJSON = readJSONSync(packageJSONPath);
  return (
    (packageJSON.devDependencies && packageJSON.devDependencies.typescript) ||
    (packageJSON.dependencies && packageJSON.dependencies.typescript)
  );
}

export const migrateCommand = new Command();

migrateCommand
  .name('migrate')
  .description('Migrate Javascript project to Typescript')
  .requiredOption('-p, --basePath <project base path>', 'Base dir path of your project.')
  .option('-e, --entrypoint <entrypoint>', 'entrypoint js file for your project')
  .option(
    '-r, --report <reportTypes>',
    'Report types separated by comma, e.g. -r json,sarif,md',
    parseCommaSeparatedList,
    []
  )
  .option('-o, --outputPath <outputPath>', 'Reports output path', '.rehearsal')
  .option('-s, --strict', 'Use strict tsconfig file')
  .option('-c, --clean', 'Clean up old JS files after TS conversion')
  .option(
    '-u, --userConfig <custom json config for migrate command>',
    'File path for custom config'
  )
  .option('-i, --interactive', 'Interactive mode to help migrate part of large apps')
  .option('-v, --verbose', 'Print more logs to debug.')
  .action(async (options: MigrateCommandOptions) => {
    const loggerLevel = options.verbose ? 'debug' : 'info';
    const logger = createLogger({
      transports: [new transports.Console({ format: format.cli(), level: loggerLevel })],
    });

    const tasks = new Listr<MigrateCommandContext>(
      [
        {
          title: 'Initialization',
          task: async (_ctx, task) => {
            // _ctx.skip = true;
            // get custom config
            const userConfig = options.userConfig
              ? new UserConfig(options.basePath, options.userConfig, 'migrate')
              : undefined;

            _ctx.userConfig = userConfig;

            const projectName = await determineProjectName(options.basePath);

            if (options.interactive) {
              const packageSelections: PackageSelection[] = discoverEmberPackages(
                options.basePath
              ).map((p) => {
                let { packageName: name } = p;
                if (name === projectName) {
                  // a little better message to indicate this option is for entire project
                  name = `${name} (this entire project)`;
                }
                return {
                  name,
                  location: p.path,
                };
              });
              const packageMap = packageSelections.reduce((map, p) => {
                map[p.name] = p.location;
                return map;
              }, {} as PackageMap);

              _ctx.input = await task.prompt<{ test: boolean; other: boolean }>([
                {
                  type: 'Select',
                  name: 'packageSelection',
                  message:
                    'We have found multiple packages in your project, select the one to migrate:',
                  choices: Object.keys(packageMap),
                },
              ]);
              // update basePath based on the selection
              _ctx.targetPacakgePath = packageMap[_ctx.input as string];
              task.title = `Initializatoin Completed! Running migration on ${_ctx.input}.`;
            } else {
              _ctx.targetPacakgePath = options.basePath;
              task.title = `Initializatoin Completed! Running migration on ${projectName}.`;
            }

            // construct migration strategy and prepare all the files needs to be migrated
            const strategy = getMigrationStrategy(_ctx.targetPacakgePath, {
              entrypoint: options.entrypoint,
              filterByPackageName: [],
            });
            const files: Array<SourceFile> = strategy.getMigrationOrder();
            _ctx.strategy = strategy;
            _ctx.sourceFilesWithAbsolutePath = files.map((f) => f.path);
            _ctx.sourceFilesWithRelativePath = files.map((f) => f.relativePath);
          },
        },
        {
          title: 'Installing dependencies',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            // install custom dependencies
            if (_ctx.userConfig?.hasDependencies) {
              task.title = `Installing custom dependencies`;
              await _ctx.userConfig.install();
            }

            if (ifHasTypescriptInDevdep(options.basePath)) {
              task.skip('typescript already exists. Skipping installing typescript.');
            } else {
              await addDep(['typescript'], true, { cwd: options.basePath });
            }
          },
        },
        {
          title: 'Creating tsconfig.json',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            if (_ctx.userConfig?.hasTsSetup) {
              task.title = `Creating tsconfig from custom config.`;
              await _ctx.userConfig.tsSetup();
            } else {
              const configPath = resolve(options.basePath, 'tsconfig.json');

              if (existsSync(configPath)) {
                task.skip(`${configPath} already exists, skipping creating tsconfig.json`);
              } else {
                task.title = `Creating ${options.strict ? 'strict' : 'basic'} tsconfig.`;

                createTSConfig(
                  options.basePath,
                  _ctx.sourceFilesWithRelativePath,
                  !!options.strict
                );
              }
            }
          },
        },
        {
          title: 'Converting JS files to TS',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            const projectName = (await determineProjectName()) || '';
            const reporter = new Reporter(projectName, options.basePath, logger);

            if (_ctx.sourceFilesWithAbsolutePath) {
              const input = {
                basePath: _ctx.targetPacakgePath,
                sourceFiles: _ctx.sourceFilesWithAbsolutePath,
                logger: options.verbose ? logger : undefined,
                reporter,
              };

              await migrate(input);

              const reportOutputPath = resolve(options.basePath, options.outputPath);
              generateReports(reporter, reportOutputPath, options.report, {
                json: jsonFormatter,
                sarif: sarifFormatter,
                md: mdFormatter,
              });
            } else {
              task.skip(
                `Skipping JS -> TS conversion task, since there is no JS file to be converted to TS.`
              );
            }
          },
        },
        {
          title: 'Clean up old JS files',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            if (_ctx.sourceFilesWithAbsolutePath.length && options.clean) {
              cleanJSFiles(_ctx.sourceFilesWithAbsolutePath);
            } else {
              task.skip(`Skipping clean up task`);
            }
          },
        },
        {
          title: 'Run Typescript compiler to check errors',
          enabled: (ctx): boolean => !ctx.skip,
          task: async () => {
            await runModuleCommand(['tsc'], { cwd: options.basePath });
          },
        },
        // TODO: what to do with those ts errors?

        {
          title: 'Creating lint config',
          enabled: (ctx): boolean => !ctx.skip,
          task: async (_ctx, task) => {
            if (_ctx.userConfig?.hasLintSetup) {
              task.title = `Creating .eslintrc.js from custom config.`;
              await _ctx.userConfig.lintSetup();
            } else {
              task.skip(`Skip creating .eslintrc.js since no custom config is provided.`);
            }
          },
        },
      ],
      { concurrent: false, exitOnError: false }
    );
    try {
      await tasks.run();
    } catch (e) {
      logger.error(`${e}`);
    }
  });

/**
 * Generate tsconfig
 */
function createTSConfig(basePath: string, fileList: string[], strict: boolean): void {
  const include = [...fileList.map((f) => f.replace('.js', '.ts'))];
  const config = strict
    ? {
        $schema: 'http://json.schemastore.org/tsconfig',
        compilerOptions: {
          baseUrl: '.',
          outDir: 'dist',
          strict: true,
          noUncheckedIndexedAccess: true,
          module: 'es2020',
          moduleResolution: 'node',
          newLine: 'lf',
          forceConsistentCasingInFileNames: true,
          noFallthroughCasesInSwitch: true,
          target: 'es2020',
          lib: ['es2020'],
          esModuleInterop: true,
          declaration: true,
          sourceMap: true,
          declarationMap: true,
        },
        include,
      }
    : {
        $schema: 'http://json.schemastore.org/tsconfig',
        compilerOptions: {
          baseUrl: '.',
          outDir: 'dist',
          emitDeclarationOnly: true,
          allowJs: true,
          target: 'es2016',
          module: 'commonjs',
          moduleResolution: 'node',
        },
        include,
      };
  writeJsonSync(resolve(basePath, 'tsconfig.json'), config, { spaces: 2 });
}

/**
 * Remove files
 * @param fileList Array of file paths
 */
function cleanJSFiles(fileList: string[]): void {
  fileList.filter((f) => f.match(/\.js$/)).forEach((f) => rmSync(f));
}
