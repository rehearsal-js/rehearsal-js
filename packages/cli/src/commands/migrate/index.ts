#!/usr/bin/env node
import { Command } from 'commander';
import { Listr } from 'listr2';
import { createLogger, format, transports } from 'winston';

import { version } from '../../../package.json';
import { parseCommaSeparatedList, gitIsRepoDirty, resetFiles } from '../../utils';
import {
  initTask,
  depInstallTask,
  convertTask,
  tsConfigTask,
  lintConfigTask,
  createScriptsTask,
} from './tasks';

import type { MigrateCommandOptions } from '../../types';

export const migrateCommand = new Command();

process.on('SIGINT', () => {
  console.log('Received SIGINT. Press Control-D to exit.');
});

migrateCommand
  .name('migrate')
  .description('migrate a javascript project to typescript')
  .option('-p, --basePath <project base path>', 'base directory of your project', process.cwd())
  .option('-e, --entrypoint <entrypoint>', 'entrypoint filepath of your project')
  .option(
    '-f, --format <format>',
    'report format separated by comma, e.g. -f json,sarif,md,sonarqube',
    parseCommaSeparatedList,
    ['sarif']
  )
  .option('-o, --outputPath <outputPath>', 'reports output directory', '.rehearsal')
  .option('-u, --userConfig <custom json config for migrate command>', 'path to rehearsal config')
  .option('-i, --interactive', 'interactive mode')
  .option('-v, --verbose', 'print debugging logs')
  .option('-d, --dryRun', 'print files that will be attempted to migrate', false)
  .action(async (options: MigrateCommandOptions) => {
    await migrate(options);
  });

async function migrate(options: MigrateCommandOptions): Promise<void> {
  const loggerLevel = options.verbose ? 'debug' : 'info';
  const logger = createLogger({
    transports: [new transports.Console({ format: format.cli(), level: loggerLevel })],
  });

  console.log(`@rehearsal/migrate ${version.trim()}`);

  if (!options.dryRun) {
    const hasUncommittedFiles = await gitIsRepoDirty(options.basePath);
    if (hasUncommittedFiles) {
      logger.warn(
        'You have uncommitted files in your repo. Please commit or stash them as Rehearsal will reset your uncommitted changes.'
      );
      process.exit(0);
    }
  }

  const defaultListrOption = {
    concurrent: false,
    exitOnError: true,
  };

  const tasks = [
    await initTask(options),
    await depInstallTask(options),
    await tsConfigTask(options),
    await lintConfigTask(options),
    await createScriptsTask(options),
  ];

  try {
    if (options.interactive) {
      // For issue #549, have to use simple renderer for the interactive edit flow
      // previous ctx is needed for the isolated convertTask
      const ctx = await new Listr(tasks, defaultListrOption).run();
      await new Listr([await convertTask(options, logger, ctx)], {
        renderer: 'simple',
        ...defaultListrOption,
      }).run();
    } else {
      await new Listr([...tasks, await convertTask(options, logger)], defaultListrOption).run();
    }
  } catch (e) {
    await resetFiles();
    logger.error(`${e}`);
  }
}
