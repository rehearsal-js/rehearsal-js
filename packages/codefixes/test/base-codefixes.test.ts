import { readdirSync, readFileSync } from 'node:fs';
import { dirname, parse, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, expect, test } from 'vitest';
import { Project } from 'fixturify-project';
import { Reporter } from '@rehearsal/reporter';
import { migrate } from '@rehearsal/migrate';
import fixturify from 'fixturify';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe('Test base codefixes', function () {
  const fixturesDir = resolve(__dirname, 'fixtures');
  const codefixesDir = resolve(__dirname, '../src/fixes/');

  // eg. ["addErrorTypeGuard","addMissingExport","addMissingTypesBasedOnInheritance","makeMemberOptional"]
  const transforms = readdirSync(codefixesDir)
    .map((file) => parse(file).name)
    .filter((item) => item.startsWith('addMissingReturnTypes')); // Exclude glint directory from test, this will be tested by glint-codefixes

  test.each(transforms)('%s', async (transform) => {
    const project = Project.fromDir(fixturesDir, { linkDeps: true, linkDevDeps: true });
    await project.write();

    const dir = 'base-codefixes';

    const files = (project.files[dir] as fixturify.DirJSON)[transform] as fixturify.DirJSON;

    expect(files, `Test files are missing for ${transform}`).toBeTypeOf('object');

    const filesToMigrate = Object.keys(files).map((file) =>
      resolve(project.baseDir, dir, transform, file)
    );

    const reporter = new Reporter({
      tsVersion: '',
      projectName: '@rehearsal/test',
      projectRootDir: project.baseDir,
    });

    const input = {
      projectRootDir: project.baseDir,
      packageDir: project.baseDir,
      filesToMigrate: filesToMigrate,
      reporter,
    };

    for await (const _ of migrate(input)) {
      // no ops
    }

    console.log(readFileSync(input.filesToMigrate[0], 'utf-8'));

    for (const file of filesToMigrate) {
      const actualOutput = readFileSync(file, 'utf-8');

      expect(actualOutput).matchSnapshot();
    }

    project.dispose();
  });
});
