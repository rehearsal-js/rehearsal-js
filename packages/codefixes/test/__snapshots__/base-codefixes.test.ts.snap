// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test base codefixes > addErrorTypeGuard 1`] = `
"// Case #1: try ... catch

const test = [\\"dummy\\"];

try {
  console.log(\\"success\\");
} catch (e) {
  console.log((e as Error).message);
  /* @ts-expect-error @rehearsal TODO TS18046: 'e' is of type 'unknown'. */
  console.log(e.notErrorMember);

  if ((e as Error).name === \\"Test\\") {
    console.log((e as Error).name);
    console.log((e as Error).name);
    console.log((e as Error).name);
    console.log((e as Error).name);
  }

  if (test[1] === undefined && (e as Error).name === \\"Test\\") {
    console.log((e as Error).name);
  }
}

// Case #2: not a try catch

function getObject<T>(): T {
  const object = [\\"dummy\\"];
  return object as unknown as T;
}

const dummy = getObject();

/* @ts-expect-error @rehearsal TODO TS18046: 'dummy' is of type 'unknown'. */
dummy.key;
"
`;

exports[`Test base codefixes > addMissingExport 1`] = `
"/* @ts-expect-error @rehearsal TODO TS2307: Cannot find module 'react' or its corresponding type declarations. */
import React from \\"react\\";

export interface Props {
  name: string;
  age: string;
}

{
  /* @ts-expect-error @rehearsal TODO TS7026: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists. */
}
const Kid = ({ name, age }: Props) => <div>{\`\${name},\${age}\`}</div>;

export default Kid;
"
`;

exports[`Test base codefixes > addMissingExport 2`] = `
"import Kid from \\"./4082-1-import\\";

export default {
  kid: Kid,
  title: \\"Kid\\",
};
"
`;

exports[`Test base codefixes > addMissingExport 3`] = `
"/* @ts-expect-error @rehearsal TODO TS2307: Cannot find module 'react' or its corresponding type declarations. */
import React from \\"react\\";

interface ToDo {
  id: string;
  title: string;
  content: string;
}

interface ToDoState {
  todos: ToDo[];
}

export interface InjectedProps {
  onAdd: () => void;
  onRemove: (id: string) => void;
  todos: ToDo[];
}

export interface AllProps extends InjectedProps {
  style: React.CSSProperties;
}

export default <P extends InjectedProps>(Component: React.ComponentType<P>) =>
  class extends React.Component<Omit<P, keyof InjectedProps>, ToDoState> {
    state = { todos: [] };

    onRemove = (id: string) => () => {
      const filteredToDos = (this.state.todos as ToDo[]).filter(
        (todo) => (todo as ToDo).id !== id
      );
      /* @ts-expect-error @rehearsal TODO TS2339: Property 'setState' does not exist on type '(Anonymous class)'. */
      this.setState({ todos: filteredToDos });
    };

    onAdd = () => {
      const todo = {
        title: \\"\\",
        content: \\"\\",
        id: \`\${Date.now}\`,
      };
      /* @ts-expect-error @rehearsal TODO TS2339: Property 'setState' does not exist on type '(Anonymous class)'. */
      this.setState((prev) => {
        [...prev.todos, todo];
      });
    };
    props: P | undefined;

    render() {
      return (
        /* @ts-expect-error @rehearsal TODO TS2307: Cannot find module 'react/jsx-runtime' or its corresponding type declarations. */
        <Component
          {...(this.props as P)}
          todos={this.state.todos}
          onAdd={this.onAdd}
          onRemove={this.onRemove}
        />
      );
    }
  };
"
`;

exports[`Test base codefixes > addMissingExport 4`] = `
"/* @ts-expect-error @rehearsal TODO TS2307: Cannot find module 'react' or its corresponding type declarations. */
import React from \\"react\\";

import makeToDo, { AllProps } from \\"./4082-2-import\\";

const ToDo = (props: AllProps) => {
  return (
{/* @ts-expect-error @rehearsal TODO TS7026: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists. */}
    <div style={props.style}>
/* @ts-expect-error @rehearsal TODO TS7006: Parameter 'todo' implicitly has an 'any' type. */
      {props.todos.map((todo) => (
{/* @ts-expect-error @rehearsal TODO TS7026: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists. */}
        <div key={todo.id}>
{/* @ts-expect-error @rehearsal TODO TS7026: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists. */}
          <div>{todo.title}</div>
{/* @ts-expect-error @rehearsal TODO TS7026: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists. */}
          <div>{todo.content}</div>
{/* @ts-expect-error @rehearsal TODO TS7026: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists. */}
          <button data-id={todo.id} onClick={() => props.onRemove(todo.id)}>
/* @ts-expect-error @rehearsal TODO TS2304: Cannot find name 'Remove'. */
            Remove
/* @ts-expect-error @rehearsal TODO TS2304: Cannot find name 'button'. */
          </button>
/* @ts-expect-error @rehearsal TODO TS2304: Cannot find name 'div'. */
        </div>
      ))}
{/* @ts-expect-error @rehearsal TODO TS7026: JSX element implicitly has type 'any' because no interface 'JSX.IntrinsicElements' exists. */}
      <button onClick={props.onAdd}>Add</button>
/* @ts-expect-error @rehearsal TODO TS2304: Cannot find name 'div'. */
    </div>
  );
};
export default makeToDo(ToDo);
"
`;

exports[`Test base codefixes > addMissingTypesBasedOnInheritance 1`] = `
"// Basics

class Animal {
  title = \\"\\";

  test(test: string): string {
    return test;
  }

  say(message: string): string {
    return message;
  }

  feed(food: Food, quantity: number): boolean {
    console.log(food, quantity);
    return false;
  }
}

interface Wild {
  wild(place: string): string;
}

interface Happy {
  happy(num: number): void;
}

class Food {
  title: string;

  constructor(title: string) {
    this.title = title;
  }
}

class Dog extends Animal {
  say(message: string): string {
    console.log(message);
    return \\"bar\\";
  }
}

export class WildDog extends Dog implements Wild, Happy {
  say(message: string): string {
    console.log(message);
    return \\"bar\\";
  }

  override feed(food: Food, quantity: number): boolean {
    console.log(food, quantity);
    return true;
  }

  wild(place: string): string {
    return place;
  }

  happy(num: number): void {
    console.log(num);
  }
}

export class Cat extends Animal {
  override say(text: string): string {
    console.log(text);
    return \\"meo\\";
  }

  feed(something: Food): boolean {
    console.log(something);
    return false;
  }
}

// Generics

type Thing = {
  foo: number;
};

// Generic Interfaces

interface GenericThingInterface<A = number, B = string> {
  thing(a: A, b: B): void;
}

interface ConstrainedThingInterface<C extends Thing> {
  thing(c: C): void;
}

export class GenericThingImp1 implements GenericThingInterface {
  thing(a: number, b: string): void {
    console.log(a, b);
  }
}

export class GenericThingImp2
  implements GenericThingInterface<boolean, number>
{
  thing(a: boolean, b: number): void {
    console.log(a, b);
  }
}

export class GenericThingImp3
  implements ConstrainedThingInterface<{ foo: number; bar: string }>
{
  thing(c: { foo: number; bar: string }): void {
    console.log(c);
  }
}

// Generic Classes

class GenericThingClass<A = number, B = string> {
  thing(a: A, b: B): void {
    console.log(a, b);
  }
}

class ConstrainedThingClass<C extends Thing> {
  thing(c: C): void {
    console.log(c);
  }
}

export class GenericThingExt1 extends GenericThingClass {
  thing(a: number, b: string): void {
    console.log(a, b);
  }
}

export class GenericThingExt2 extends GenericThingClass<boolean, number> {
  thing(a: boolean, b: number): void {
    console.log(a, b);
  }
}

export class GenericThingExt3 extends ConstrainedThingClass<{
  foo: number;
  bar: string;
}> {
  thing(c: { foo: number; bar: string }): void {
    console.log(c);
  }
}

// Multiple signatures

interface MultipleInterface {
  foo(a: number): number;
  foo(b: string): string;
  foo(a: number, b: string): boolean;

  moo(a: string): string;
}

class MultipleClass {
  boo(a: number): number;
  boo(b: string): string;
  boo(a: number, b: string): boolean;
  boo(ab: number | string, b?: string): number | string | boolean {
    return !ab || !b;
  }
}

export class MultipleImpl extends MultipleClass implements MultipleInterface {
  /* @ts-expect-error @rehearsal TODO TS2416: Property 'boo' in type 'MultipleImpl' is not assignable to the same property in base type 'MultipleClass'..  Type '(ab: any, b?: any) => boolean' is not assignable to type '{ (a: number): number; (b: string): string; (a: number, b: string): boolean; }'..    Type 'boolean' is not assignable to type 'number'. */
  boo(ab, b?): boolean {
    return !ab || !b;
  }

  /* @ts-expect-error @rehearsal TODO TS2416: Property 'foo' in type 'MultipleImpl' is not assignable to the same property in base type 'MultipleInterface'..  Type '(ab: any, b?: any) => boolean' is not assignable to type '{ (a: number): number; (b: string): string; (a: number, b: string): boolean; }'..    Type 'boolean' is not assignable to type 'number'. */
  foo(ab, b?): boolean {
    return !ab || !b;
  }

  moo(b: string): string {
    return b;
  }
}
"
`;

exports[`Test base codefixes > addMissingTypesBasedOnInlayHints 1`] = `
"export function test1(): string {
  return \\"Test this\\";
}

export const test2 = () => console.log(\\"Test this\\");

export const test3 = () => \\"Test this\\";

export function test4<Type>(arg: Array<Type>): Type[] {
  return arg;
}

export async function test5(): Promise<TestClass> {
  return new TestClass();
}

interface TestInterface {
  test1(): string | undefined;
}

class BaseTestClass {
  test2(param: boolean): number | string {
    return param ? 0 : \\"0\\";
  }
}

export class TestClass extends BaseTestClass implements TestInterface {
  test1(): string | undefined {
    return undefined;
  }

  test2(): string | number {
    return 0;
  }
}
"
`;

exports[`Test base codefixes > annotateWithStrictTypeFromJSDoc 1`] = `
"export class Foo {
  value = 0;
}

export class GenericClass<something = number, somethingElse = number> {
  thing(a: something, b: somethingElse): void {
    console.log(a, b);
  }
}

/** @type {number | string} */
export const a: number | string = 0;

/** @type {number | string | UnavailableType} */
export const b = 0;

/** @type {GenericClass<string, string, string>} */
export const c = 0;

export class ThinkClass {
  /** @type {number | null} */
  a: number | null = 0;

  /** @type {number | UnavailableType} */
  b = 0;

  /**
   * JSDoc
   *
   * @param {number} a                                //
   * @param {(string|number)} b                       // ParenthesizedType.type => UnionType
   * @param {Foo} c                                   // TypeReference
   * @param {number|UnavailableType} d                // UnionType
   * @param {() => boolean} e                         // FunctionType
   * @param {GenericClass<string, string>} f          // TypeReference

   * @param {<callbackFunction>} g                    // FunctionType

   * @param {UnavailableType} h                       // TypeReference
   * @param {Foo|UnavailableType} i                   // UnionType
   * @param {() => UnavailableType} j                 // FunctionType
   * @param {GenericClass<string, UnavailableType>} k // TypeReference
   * @param {number|() => UnavailableType} l          // UnionType
   * @param {(a: UnavailableType) => void} m          // FunctionType
   * @param {(a, b: string) => void} n                // FunctionType

   * @param {GenericClass<string, string, string>} o  // TypeReference... 3 args instead of 2

   * @param {{ a: string, b: Date }} p
   * @param {{ a: string, b: UnavailableType }} q
   * @param {function(this:{ a: string}, string, number): boolean} r

   * @return {void}
   */
  more(
    a: number,
    b: string | number,
    c: Foo,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'd' implicitly has an 'any' type. */
    d,
    e: () => boolean,
    f: GenericClass<string, string>,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'g' implicitly has an 'any' type. */
    g,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'h' implicitly has an 'any' type. */
    h,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'i' implicitly has an 'any' type. */
    i,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'j' implicitly has an 'any' type. */
    j,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'k' implicitly has an 'any' type. */
    k,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'l' implicitly has an 'any' type. */
    l,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'm' implicitly has an 'any' type. */
    m,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'a' implicitly has an 'any' type. */
    n: (a, b: string) => void,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'o' implicitly has an 'any' type. */
    o,
    p: { a: string; b: Date },
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'q' implicitly has an 'any' type. */
    q,
    r: (this: { a: string }, arg1: string, arg2: number) => boolean,
    /* @ts-expect-error @rehearsal TODO TS7006: Parameter 's' implicitly has an 'any' type. */
    s
  ): void {
    return console.log(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s);
  }
}

/**
 * JSDoc
 *
 * @param {number} a                                //
 * @param {(string|number)} b                       // ParenthesizedType.type => UnionType
 * @param {Foo} c                                   // TypeReference
 * @param {number|UnavailableType} d                // UnionType
 * @param {() => boolean} e                         // FunctionType
 * @param {GenericClass<string, string>} f          // TypeReference

 * @param {<callbackFunction>} g                    // FunctionType

 * @param {UnavailableType} h                       // TypeReference
 * @param {Foo|UnavailableType} i                   // UnionType
 * @param {() => UnavailableType} j                 // FunctionType
 * @param {GenericClass<string, UnavailableType>} k // TypeReference
 * @param {number|() => UnavailableType} l          // UnionType
 * @param {(a: UnavailableType) => void} m          // FunctionType
 * @param {(a, b: string) => void} n                // FunctionType

 * @param {GenericClass<string, string, string>} o  // TypeReference... 3 args instead of 2

 * @param {{ a: string, b: Date }} p                //
 * @param {{ a: string, b: UnavailableType }} q     //
 * @param {function(this:{ a: string}, string, number): boolean} r

 * @return {void}
 */
export function think(
  a: number,
  b: string | number,
  c: Foo,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'd' implicitly has an 'any' type. */
  d,
  e: () => boolean,
  f: GenericClass<string, string>,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'g' implicitly has an 'any' type. */
  g,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'h' implicitly has an 'any' type. */
  h,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'i' implicitly has an 'any' type. */
  i,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'j' implicitly has an 'any' type. */
  j,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'k' implicitly has an 'any' type. */
  k,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'l' implicitly has an 'any' type. */
  l,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'm' implicitly has an 'any' type. */
  m,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'a' implicitly has an 'any' type. */
  n: (a, b: string) => void,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'o' implicitly has an 'any' type. */
  o,
  p: { a: string; b: Date },
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 'q' implicitly has an 'any' type. */
  q,
  r: (this: { a: string }, arg1: string, arg2: number) => boolean,
  /* @ts-expect-error @rehearsal TODO TS7006: Parameter 's' implicitly has an 'any' type. */
  s
): void {
  return console.log(a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s);
}
"
`;

exports[`Test base codefixes > makeMemberOptional 1`] = `
"export interface Animal {
  species: string;
  weight?: number;
}

export interface Car {
  make: string;
  model: string;
  color?: string;
}
"
`;

exports[`Test base codefixes > makeMemberOptional 2`] = `
"export class Employee {
  name: string;
  badgeNumber?: string;

  constructor({ name, badgeNumber }: { name: string; badgeNumber: string }) {
    this.name = name;
    this.badgeNumber = badgeNumber;
  }
}
"
`;

exports[`Test base codefixes > makeMemberOptional 3`] = `
"export interface Rabbit {
  color?: string;
  species: string;
  name: string;
}

export function feedRabbit(rabbit: Rabbit): void {
  console.log(\`Rabbit \${rabbit.name} loves to eat fruit, not hay.\`);
}
"
`;

exports[`Test base codefixes > makeMemberOptional 4`] = `
"export default interface Human {
  height?: number;
  weight: number;
}
"
`;

exports[`Test base codefixes > makeMemberOptional 5`] = `
"import { Animal, Car } from \\"./2790-import-1\\";
import { Employee } from \\"./2790-import-2\\";
import * as RabbitStuff from \\"./2790-import-3\\";
import Human from \\"./2790-import-4\\";

const animal: Animal = {
  species: \\"dog\\",
  weight: 90,
};

delete animal.weight;

const car: Car = {
  make: \\"Ford\\",
  model: \\"Focus\\",
  color: \\"red\\",
};

delete car.color;

const employee = new Employee({ name: \\"Victor\\", badgeNumber: \\"12345\\" });
delete employee.badgeNumber;

const human: Human = {
  weight: 200,
  height: 180,
};
delete human.height;

const rabbit: RabbitStuff.Rabbit = {
  color: \\"butterscotch\\",
  species: \\"Holland Lop\\",
  name: \\"Poca\\",
};
delete rabbit.color;

interface Person {
  name: string;
  age?: number;
  address?: string | null;
}

const person: Person = {
  name: \\"Jane\\",
  age: 10,
  address: \\"123 Happy St\\",
};

delete person.address;

class Student {
  name: string;
  grade: number;
  gender?: string;

  constructor({
    name,
    grade,
    gender,
  }: {
    name: string;
    grade: number;
    gender: string;
  }) {
    this.name = name;
    this.grade = grade;
    this.gender = gender;
  }
}
const student = new Student({ name: \\"\\", grade: 3, gender: \\"male\\" });
delete student.gender;

type Tree = {
  species: string;
  age: number;
  height?: number;
};

const oak: Tree = {
  species: \\"oak\\",
  age: 150,
  height: 40,
};
delete oak.height;
"
`;
